File: Web/ftp-db/frontend/src/actions/retrieve.ts
"use server";

import { OpaqueError, Result } from "@/lib/utils";

type Response = {
  content: string;
};

export async function retrieveNote(
  id: string,
): Promise<Result<string, string>> {
  const response = await fetch(`${process.env.API_URL}/retrieve/${id}`);
  if (!response.ok) {
    const { error } = (await response.json()) as OpaqueError;
    return {
      ok: false,
      error,
    };
  }
  const { content } = (await response.json()) as Response;
  return {
    ok: true,
    result: content,
  };
}

File: Web/ftp-db/backend/src/routes/retrieve.rs
use crate::error::{opaque_error, OpaqueError};
use crate::ftp::get_ftp;
use rocket::http::Status;
use rocket::response::status;
use rocket::serde::json::Json;
use rocket::State;
use serde::Serialize;
use crate::state::AppState;

#[derive(Serialize)]
pub struct RetrieveResponse {
    content: String,
}

#[get("/retrieve/<id>")]
pub async fn retrieve(state: &State<AppState>, id: &str) -> Result<Json<RetrieveResponse>, status::Custom<Json<OpaqueError>>> {
    let mut ftp = get_ftp();
    let files = state.files.lock().unwrap();
    let filename = files.get(id);
    let Some(filename) = filename else {
        return Err(opaque_error("", Status::BadRequest, "File not found"));
    };
    let cursor = ftp.stream.simple_retr(filename).map_err(|error| opaque_error(error, Status::BadRequest, "Could not retrieve file"))?;
    let buf = cursor.into_inner();
    let content = str::from_utf8(&buf).map_err(|error| opaque_error(error, Status::InternalServerError, "Could not parse file"))?;

    Ok(Json(RetrieveResponse {
        content: content.to_owned(),
    }))
}